KAFKA-13306: Null connector config value passes validation, but fails creation

lhunyady opened the pull request:
This patch adds null value check to the connector config validation, and extends unit tests to cover this functionality.

Commit:
KAFKA-13306: Null connector config value passes validation, but fails creation

This patch adds null value check to the connector config validation, and extends unit tests to cover this functionality.

Co-authored-by: Chris Egerton <chrise@confluent.io>

abbccdda commented on connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java:
```diff
@@ -150,4 +152,15 @@ public Config validate(Map<String, String> connectorConfigs) {
+                .collect(Collectors.joining(", "));
+
+        if (!keysWithNullValue.isEmpty()) {
+            throw new ConnectException(String.format("Null value found in config for key(s) %s", keysWithNullValue));
```
Should this just be IllegalArgumentException?

lhunyady replied:
Thank you for the comment, updated it in the latest commit.

abbccdda commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/AbstractHerderTest.java:
```diff
@@ -289,6 +290,23 @@ public void testBuildRestartPlanForUnknownConnector() {
     }
 
+    @Test
+    public void testConfigValidationNullConfig() {
```
Could you also add a test case with multiple null values as well?

lhunyady replied:
Thank you for the comment, updated it in the latest commit.

lhunyady commented:
@C0urante  @kkonstantine Can you please review this change for me?

lhunyady commented:
Thanks @C0urante for the feedback, I agree with you and updated the PR accordingly.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java:
```diff
@@ -576,8 +576,12 @@ private static ConfigInfos validateClientOverrides(String connName,
     }
 
-    // public for testing
     public static ConfigInfos generateResult(String connType, Map<String, ConfigKey> configKeys, List<ConfigValue> configValues, List<String> groups) {
+        return generateResult(connType, configKeys, configValues, groups, Collections.emptyMap());
+    }
```
Since this is no longer used in the main code base, could we remove it from this class and, where necessary, replace it either with a call to the new variant created below, or a testing-only wrapper function that has the same body as this one?

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/rest/entities/ConfigInfo.java:
```diff
@@ -60,6 +60,6 @@ public int hashCode() {
     @Override
     public String toString() {
-        return "[" + configKey.toString() + "," + configValue.toString() + "]";
+        return "[" + configKey + "," + configValue + "]";
```
Nice cleanup 👍

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java:
```diff
@@ -603,9 +607,39 @@ public static ConfigInfos generateResult(String connType, Map<String, ConfigKey>
+    private static List<ConfigValueInfo> buildConfigValueInfoWithNullValues(Map<String, String> connectorProps) {
+        return connectorProps.entrySet().stream()
+                .filter(configEntry -> configEntry.getValue() == null)
+                .map(configEntry -> new ConfigValueInfo(configEntry.getKey(), String.format("Null Value Not Allowed for key %s.", configEntry.getKey())))
```
I believe that including the name of the property in the error message is redundant as that information will be available already in the REST response.

I also think we may want to be clearer about the error message here. Users can't supply null values, but developers (by specifying `null` as the default value for a property in a `ConfigDef`, for example) definitely can, and we may want to make it clear which variety we're prohibiting.

What do you think about this?
```suggestion
                .map(configEntry -> new ConfigValueInfo(configEntry.getKey(), "The JSON literal `null` may not be used in connector configurations"))
```

C0urante reviewed (commented):
Thanks @lhunyady. I left a few comments but I was also wondering about an alternative approach that seems a little less verbose; I've drafted it on a personal branch (including the tests you've added in this PR, which both pass); what do you think?

https://github.com/C0urante/kafka/blob/e59b50a369390f8c6645103989e6eb17273931a3/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractHerder.java#L445-L451

Also, higher-level, I'm wondering if it might be easiest to just filter out `null` values from `connectorProps` instead of failing the connector? I'm a little hesitant to do that just because people might become reliant on this behavior but it's probably worth considering before we merge this PR.

Commit:
KAFKA-13306: Null connector config value passes validation, but fails creation

This patch adds null value check to the connector config validation, and extends unit tests to cover this functionality.

Co-authored-by: Chris Egerton <chrise@confluent.io>

lhunyady commented:
Thank you @C0urante, I have updated the pull request with your valuable suggestions.

C0urante reviewed (approved):
LGTM, thanks @lhunyady!

lhunyady commented:
@C0urante what are the next steps to get this pull request merged?

C0urante commented:
@tombentley @kkonstantine would either of you have time for this small but valuable fix?

mimaison reviewed (approved):
LGTM, thanks for the PR
